### PR TITLE
[Perf] Optimize Qwen3-VL unique-image serving on H100

### PR DESCRIPTION
**Pull request by BBuf**
## Motivation

The target traffic is one-pass streaming: an image is normally processed once and is not seen again. The previous cache-hot version automatically reserved 16 GiB for preprocessing artifacts plus 8 GiB for multimodal embeddings, which wasted memory and hid preprocessing, feature-transport, and scheduling costs.

This revision keeps preprocessing and embedding retention disabled for unique-image traffic, retains the request-local Qwen3-VL fast path, and removes the remaining per-image CUDA IPC reconstruction copy. After the review fixes and current-main merge, the refreshed three-run median is **68.551 requests/s**. The historical vLLM median under the same workload is **61.657 requests/s**, so the refreshed SGLang result is **+11.18%**.

## What changed

- Build request-local Qwen3-VL image artifacts without preprocessing-cache lookup, single-flight state, or retention. The fast path directly produces prompt IDs, image offsets, and M-RoPE metadata.
- Cache only normalized chat-template rendering/tokenization metadata in a bounded 128-entry LRU.
- Skip raw prompt tokenization when the Qwen-VL processor produces the final expanded input IDs.
- Split the reusable prompt KV boundary from the generated suffix and prioritize the prompt node in radix eviction.
- Keep high-concurrency decode inside CUDA graph, use stock FlashInfer decode/page size 1, and use the profiled prefill/decode interval 22.
- Reserve eager-prefill activation headroom independently of embedding-cache retention.
- Defer CUDA IPC feature reconstruction for single-GPU Qwen3-VL. When an entire image is consumed by the current prefill chunk, the ViT borrows a zero-copy view of the bounded IPC pool and acknowledges it on the CUDA stream after reads are enqueued. Multi-GPU, cross-chunk, cache-hit, uncached-handle, and failure paths keep the existing copy/ack behavior.
- For the profiled large-Hopper configuration (`max_running_requests >= 400`, no preprocessing/embedding retention), increase the bounded IPC pool from 1 GiB to 3 GiB.

Automatic serving defaults remain restricted to Qwen3-VL/Qwen3-VL-MoE on SM90 GPUs with at least 60 GiB. Explicit cache, transport, attention, interval, CUDA-graph, IPC-pool, and radix-eviction choices remain authoritative.

## Review follow-up fixes

The branch now includes the following merge-readiness fixes and regressions:

- Mark `prefill_decode_interval`, `radix_eviction_policy`, `mm_preprocess_cache_size_mb`, and `mm_feature_transport` as valid model-resolvable fields. This fixes the startup failure reported by DoorDash.
- Preserve an explicit `--radix-eviction-policy lru`, including CLI `--flag=value` and YAML configuration; `priority` is applied only when the option is omitted.
- Do not load Hugging Face model configuration from the decode-graph memory hook when GPU-memory information is unavailable.
- Add coverage for the Qwen3-VL override contract, chat-template cache isolation, request-local media artifacts, CUDA IPC one-shot borrowing, and prompt-prefix KV retention in both legacy and unified radix caches.

The final diff against `main` modifies 18 implementation files and 7 test files.

### Environment

- Final PR head: `00a555c5dadd11fbc6b27842dd2e896898a75060`
- Merged `main`: `12735c2d7684647454efb033a143cad92bfd305a`
- GPU: 1x NVIDIA H100 80GB HBM3, TP=1
- Model: `Qwen/Qwen3-VL-4B-Instruct`
- Model revision: `ebb281ec70b05090aa6165b016eac8ec08e71b17`
- Validation base image: `lmsysorg/sglang@sha256:4c44c95626c7601f19259266342341976e848f835fbe33ebb4b09a28dff3063c`
- Dependency overlay: FlashInfer Python/cubins 0.6.18, JIT cache 0.6.18+cu130, CUTLASS 4.6.2, QuACK 0.6.4

The candidate source was bind-mounted into the validation container. The dependency overlay is recorded here explicitly; the CI image from this pushed head should be used as the single-digest external reproduction image.

```bash
CUDA_VISIBLE_DEVICES=3 \
taskset -c 0-31,64-95 \
python3 -m sglang.launch_server \
  --model-path Qwen/Qwen3-VL-4B-Instruct \
  --revision ebb281ec70b05090aa6165b016eac8ec08e71b17 \
  --host 127.0.0.1 --port 31000 \
  --enable-metrics \
  --max-running-requests 400 \
  --chunked-prefill-size 65536 \
  --mem-fraction-static 0.95 \
  --context-length 4096 \
  --tp-size 1
```

Resolved Qwen3-VL/Hopper defaults were preprocessing cache 0 MiB, CUDA IPC feature transport with a 3 GiB pool, priority radix eviction, prefill/decode interval 22, FA3 prefill, FlashInfer decode, and decode CUDA graph max batch 400.

### Workload

- deterministic performance-equivalent surrogate with 1,200 distinct synthetic shelf-tag JPEGs plus 16 separate warm-up images;
- 600 images at 512x480 and 600 at 512x512, median 248 merged vision tokens;
- manifest SHA-256: `d92eebbe2ac9bc3421e68ac09a1db3e9c60c1a7faff8e209ef8599e32e100c8f`;
- 5,178-character shared text prefix before the image;
- streaming OpenAI chat, concurrency 373;
- `max_completion_tokens=256`, `temperature=0.1`, `repetition_penalty=1.1`, EOS enabled;
- fresh server and 16-image warm-up before each measured stream;
- GPU process snapshots empty before launch and after shutdown.

This surrogate matches the documented vision shapes but is not DoorDash's private production corpus.

## End-to-end performance

| Engine / snapshot | Requests/s | Wall time | Completed / fields present | Median TTFT | Output tokens |
|---|---:|---:|---:|---:|---:|
| Historical SGLang before PR | 57.090 | 21.020 s | 1,200 / 1,200 | 2.620 s | 128,268 |
| Historical PR before one-shot IPC borrowing | 66.104 median | 18.153 s | 1,200 / 1,200 | 1.323 s | — |
| Historical vLLM median | 61.657 | 19.463 s | 1,200 / 1,200 | 0.378 s | — |
| Original PR report median | 69.925 | 17.161 s | 1,200 / 1,200 | 1.304 s | — |
| Refreshed SGLang run 1 | 68.143 | 17.610 s | 1,200 / 1,200 | 1.301 s | 128,264 |
| Refreshed SGLang run 2 | 68.551 | 17.505 s | 1,200 / 1,200 | 1.310 s | 128,261 |
| Refreshed SGLang run 3 | 68.702 | 17.467 s | 1,200 / 1,200 | 1.323 s | 128,262 |
| **Refreshed median** | **68.551** | **17.505 s** | **1,200 / 1,200** | **1.310 s** | — |
| Final-head confirmation after last main merge | 68.379 | 17.549 s | 1,200 / 1,200 | 1.297 s | 128,266 |

The refreshed three-run mean is 68.465 requests/s, sample standard deviation is 0.289 requests/s, and CV is 0.422%. The final-head confirmation is inside that range. The refreshed median is 1.97% below the original 69.925 report and 11.18% above the historical 61.657 vLLM median. vLLM was not rerun during this review-fix pass.

## Semantic accuracy

The manifest provides ground truth for nine parsed fields. The final-head confirmation produced:

| Metric over 1,200 requests | Final head |
|---|---:|
| All fields present | 1,200 |
| All 9 fields exact | 639 |
| Variant + regular/sale/loyalty/unit prices exact | 1,200 |
| MSID exact | 1,189 |
| Barcode exact | 1,186 |
| Descriptive name exact | 1,065 |
| Tag name exact | 648 |
| Median completion tokens | 105 |

The three refreshed runs contained 640, 640, and 643 all-nine-field exact matches. Sampling causes small movement in free-text fields, while all deterministic variant and price fields remain exact. This is within or slightly above the earlier PR accuracy range.

## MFU and profile evidence

For the median run, an operator-level count of the vision encoder/merger, uncached text prefill, decode projections, LM head, and dense attention gives 4.599 PFLOP over 17.505 seconds: **262.7 TFLOP/s, approximately 26.6% of the dense H100 SXM BF16 peak**. The estimate accounts for shared-prefix KV reuse and excludes elementwise/host work, so it is useful model MFU rather than a hardware-counter FLOP rate. Active one-second telemetry averaged 97.2% SM activity, 47.6% memory activity, and 617.8 W.

Stage-separated torch traces show BF16 nvJet GEMMs consuming about 71% of decode and 70% of extend device time. FlashInfer decode attention is about 10%; FA3 extend attention is about 8.5%.

Focused Nsight Compute evidence on the exact gate-up projection:

| Shape | Throughput | Applicable-roof efficiency | Key limit |
|---|---:|---:|---|
| `M=336, K=2560, N=19456` | 542.2 TFLOP/s | 55.3% | one wave/SM, register/shared-memory pressure, shared-store conflicts |
| `M=8, K=2560, N=19456` | 20.0 TFLOP/s | 74.8% | bandwidth-bound, 2.58 TB/s, only 76 blocks / 0.58 wave per SM |

The most evidence-backed follow-up is a separate Hopper BF16 small-M gate-up path for M=1/2/4/8 that increases weight-streaming parallelism. Image caching is intentionally unavailable for this one-pass workload, and attention is not the dominant device-time consumer.

## Validation

```text
test_model_overrides.py: 94 passed
test_server_args.py: 192 passed
test_serving_chat.py: 123 passed
multimodal preprocess-cache: 27 passed, 1 skipped
multimodal GPU feature transport: 37 passed
H100 CUDA IPC transport: 6 passed
media artifact, feature materialization, async processor, abort cleanup: passed
legacy + unified radix prompt-retention tests (page sizes 1/4/16): passed
pre-commit, AST, ruff, format, codespell, CI registry validation: passed
H100 final-head stream: 1,200/1,200 completed, fields present, no competing GPU process
```

## Residual risk

- The zero-copy path is intentionally limited to TP=1 and an image fully consumed by the current chunk. Multi-GPU and cross-chunk inputs retain the old reconstruction copy.
- The 3 GiB bounded IPC pool trades approximately 4.6% KV capacity for avoiding CPU fallback in the profiled 400-request wave. It is restricted to the no-retention Hopper profile and can be overridden explicitly.
- The prompt/output radix split affects all model families and should receive a radix-cache-owner review in addition to the existing unit coverage.
- Public VLM accuracy suites, multi-GPU configurations, and a single-digest CI image remain useful follow-ups.

## Checklist

- [x] Unique images; no measured image reuse.
- [x] Shared text-prefix KV is the only intentional cross-request reuse.
- [x] Startup failure reported by the external reproducer fixed and covered.
- [x] Semantic accuracy, stage profiler, workload MFU, and NCU evidence collected.
- [x] Latest `main` merged and final-head H100 confirmation completed.
- [x] No custom CUDA or FlashInfer kernel added by this PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #34668254311](https://github.com/sgl-project/sglang/actions/runs/34668254311)
Latest PR Test (Extra): :x: [Run #34668254036](https://github.com/sgl-project/sglang/actions/runs/34668254036)
Latest PR Test (AMD ROCm 10): :x: [Run #34668254415](https://github.com/sgl-project/sglang/actions/runs/34668254415)

### CODE REVIEW

**Comment by yhyang201** [comment]
Changes look correct to me overall, but it would be good to get a review of the multimodal design from @mickqian, and have someone familiar with the radix cache review those changes as well, since the prompt/output split affects all models, not just Qwen3-VL.

**Review by yhyang201 — APPROVED** [review]

**Comment by chuqiwang** [comment]
Thanks for this — following up on the repro, since our benchmark setup is referenced here.
We tried reproducing the numbers using this exact commit (`522a1f41a`) and the documented command/environment, but hit a startup failure before we could run the throughput comparison:
```
ValueError: _qwen3vl_hopper_serving_overrides: [...] not model-overridable
```
Digging in, it looks like the automatic Hopper serving-default overrides in `qwen3_vl.py` reference four fields from `server_args.py`:
- `prefill_decode_interval`
- `radix_eviction_policy`
- `mm_preprocess_cache_size_mb`
- `mm_feature_transport`

These fields aren't marked `resolvable=True` in this commit, so the whitelist check in `overrides.py::validate_declarations()` rejects them at startup.
Is this expected on our end, or could these four fields be missing `resolvable=True` in this PR? We're happy to share the exact diff we applied locally to unblock our repro if useful—in case this is a quick fix rather than something intentional we're misunderstanding.

**Comment by BBuf** [comment]
@chuqiwang Thanks for the detailed repro — your diagnosis was correct. The four fields were missing from the model-override whitelist.

I pushed the fix in `429ff742b1` and added a golden test that validates the complete Qwen3-VL/Hopper declaration. The current PR head is `00a555c5dadd`, merged with `main` through `12735c2d768`. I also fixed two issues found while validating the update:

- an explicit `--radix-eviction-policy lru` (CLI or YAML) is now preserved;
- decode-graph expansion no longer attempts to load model configuration when GPU-memory information is unavailable.

The relevant CPU suites now pass (`test_model_overrides.py`: 94/94, `test_server_args.py`: 192/192), and the server starts successfully on H100 with the intended resolved defaults.

I repeated the 1,200-distinct-image stream at concurrency 373 on an isolated H100:

| Run | Requests/s | Completed | All fields present |
|---:|---:|---:|---:|
| 1 | 68.143 | 1200/1200 | 1200/1200 |
| 2 | 68.551 | 1200/1200 | 1200/1200 |
| 3 | 68.702 | 1200/1200 | 1200/1200 |

Median is **68.551 requests/s** with 0.42% CV. After the last three `main` commits were merged, a final-head confirmation produced **68.379 requests/s**, also 1200/1200. The manifest SHA-256 is `d92eebbe2ac9bc3421e68ac09a1db3e9c60c1a7faff8e209ef8599e32e100c8f`.

For the MFU follow-up: the median run performs an estimated 4.599 PFLOP of useful vision, uncached-prefill, decode, LM-head, and attention work in 17.505 seconds: **262.7 TFLOP/s, approximately 26.6% of the dense H100 SXM BF16 peak**. Active telemetry averaged 97.2% SM, 47.6% memory, and 617.8 W. Torch profiler attributes roughly 70–71% of extend/decode device time to BF16 GEMMs.

Focused NCU on the gate-up projection measured 542.2 TFLOP/s at `M=336`, while the `M=8` tail is bandwidth/grid limited at 20.0 TFLOP/s (2.58 TB/s, 76 blocks for 132 SMs). This points to a Hopper BF16 small-M MLP path as the next optimization, rather than additional image caching.

I updated the PR body with the exact launch configuration, semantic-accuracy results, test coverage, MFU methodology, and NCU breakdown. Once the new-head CI image is available, we can exchange that single digest and compare resolved startup arguments/NUMA placement against the 50.27-QPS deployment.

**Review by mickqian — COMMENTED** [review]
Three correctness/startup issues to address before enabling these defaults:

File: python/sglang/srt/mem_cache/kv_cache_configurator.py [hunk 1]
```diff
@@ -139,21 +149,29 @@ def _should_enable_lazy_compaction() -> bool:
...
+    """Reserve persistent multimodal allocations and eager-prefill activations."""
     if not is_multimodal:
         return 0.0
-    reserved_mb = envs.SGLANG_VLM_CACHE_SIZE_MB.get()
+    activation_reserve_mb = (
+        _MM_RUNTIME_ACTIVATION_RESERVE_MB
+        if model_type in _MM_RUNTIME_ACTIVATION_RESERVE_MODEL_TYPES
+        else 0
+    )
+    reserved_mb = envs.SGLANG_VLM_CACHE_SIZE_MB.get() + activation_reserve_mb
```
Comment (mickqian) [inline python/sglang/srt/mem_cache/kv_cache_configurator.py]: What exactly does this extra fixed 6 GiB cover beyond the activation/runtime headroom already reserved through `mem_fraction_static`? Applying it unconditionally to every Qwen3-VL variant can make otherwise-startable configurations fail: [current NVIDIA CI](https://github.com/sgl-project/sglang/actions/runs/33949480150/job/101596030877) already fails Qwen3-VL-30B-A3B startup with no KV budget after reporting 19.29 GiB free and a 6.50 GiB multimodal reservation, so this needs to be sized jointly with existing headroom rather than added as a blanket reserve.

File: python/sglang/srt/models/qwen3_vl.py [hunk 2]
```diff
@@ -1427,13 +1431,29 @@ def _get_visual_feature(
...
+            items, range(len(items)), borrow_one_shot=True
+        )
         assert pixel_values.dim() == 2, pixel_values.dim()
-        return self.visual(pixel_values, grid_thw=grid_thw)
+        try:
+            return self.visual(pixel_values, grid_thw=grid_thw)
+        finally:
+            for item, proxy in borrowed_features:
+                proxy.release_borrowed_on_current_stream()
+                item.feature = None
```
Comment (mickqian) [inline python/sglang/srt/models/qwen3_vl.py]: An image fitting entirely in the current chunk does not guarantee one-time use: retracted requests retain these multimodal items and may need to run ViT again, but this clears their input feature while the new Hopper defaults disable the embedding cache. Please retain a request-owned input or a guaranteed recomputation path until it is no longer needed, and add a retract/re-prefill regression test; a local control-flow repro with the CUDA boundary mocked currently fails on the second encode with `expected torch.Tensor, got NoneType`.

File: python/sglang/srt/arg_groups/model_overrides/qwen3_vl.py [hunk 3]
```diff
@@ -30,3 +79,61 @@ def _qwen3vl_overrides(server_args: Any, hf_config: Any) -> dict:
...
+    updates = {}
+    preprocess_cache_size_mb = cfg.mm_preprocess_cache_size_mb
+    if preprocess_cache_size_mb is None:
+        preprocess_cache_size_mb = 0
+        updates["mm_preprocess_cache_size_mb"] = 0
+    cache_retention_enabled = (
+        preprocess_cache_size_mb > 0 or envs.SGLANG_VLM_CACHE_SIZE_MB.get() > 0
+    )
+    if cfg.mm_feature_transport is None and not cache_retention_enabled:
+        updates["mm_feature_transport"] = "cuda_ipc"
```
Comment (mickqian) [inline python/sglang/srt/arg_groups/model_overrides/qwen3_vl.py:113]: This also selects `cuda_ipc` when `nnodes > 1`, so a two-node Hopper Qwen3-VL deployment with transport left unset now fails in `handle_multimodal_feature_transport()` with `--mm-feature-transport=cuda_ipc only supports a single node`. Please gate automatic IPC on a supported single-node topology and add a multi-node default-resolution test.

Comment (BBuf) [reply python/sglang/srt/arg_groups/model_overrides/qwen3_vl.py:113]: Fixed: automatic cuda_ipc is now gated by nnodes == 1, and the 3 GiB pool override only applies when the resolved transport is actually cuda_ipc. Added a multi-node resolution test. Fixed in 031ed82d8b.

File: python/sglang/srt/mem_cache/kv_cache_configurator.py (hunk 1 above)
Comment (BBuf) [reply python/sglang/srt/mem_cache/kv_cache_configurator.py]: Removed the fixed 6 GiB reserve. Qwen3-VL now uses the existing mem_fraction_static activation/runtime headroom; mm_runtime_reservation_gb only accounts for persistent embedding and GPU transport pools. Fixed in 031ed82d8b.

File: python/sglang/srt/models/qwen3_vl.py (hunk 2 above)
Comment (BBuf) [reply python/sglang/srt/models/qwen3_vl.py]: Fixed by dropping the one-shot borrow path. The first materialization now retains a request-owned tensor, so a retracted request can run ViT again. Added test_retract_reprefill_retains_request_owned_visual_input. Fixed in 031ed82d8b.

Comment (BBuf) [reply python/sglang/srt/models/qwen3_vl.py]: Fixed in 77ba4006c8. The first ViT pass releases the IPC lease after the packed input copy, then preserves a pinned-host copy on a side stream for re-prefill. Added re-prefill and lease-lifetime tests. A forced-retraction H100 run completed 8/8 requests; the 3-run c373 median is 67.11 req/s vs 67.32 before, with 0 failures.

**Review by mickqian — APPROVED** [review]
